SECURITY.WHITEPAPER.md

```diff
@@ -0,0 +1,113 @@
+# Datasafe Security Whitepaper
+
+## General information
+Datasafe is a flexible encryption library. It uses different encryption algorithms. Some of them can be 
+configured by client application. It uses encryption algorithms provided by Bouncy Castle library by default.
+CMS (crypto message syntax) standard [RFC5652](https://tools.ietf.org/html/rfc5652.html) employed for storing private 
+files encrypted with symmetric keys as well as for sharing files with other users using asymmetric key pairs. 
+
+## Default storage configuration
+Datasafe library can operate with different type of storages: filesystem, datasabe, s3 compatible(amazon aws, ceph, minio).
+Other storage types can be added by implementing storage api interface. 
+Each library user can use more then one storage. All user storages form users dfs (distributed file system)
+By default on system start have to be configured one system dfs where users profiles will be stored.
+New user registers in datasafe directory service using his name and password.
+Default user creation process consists of creation private profile, public profile, keystore with public keys, keystore with private keys.
+```
+Default location within system dfs:
+    /profiles
+        /private
+            /username - user's private profile 
+        /public
+            /username - user's public profile   
+    /users
+        /public
+            /pubkeys - keystore consists user's public key
+            /inbox - location of shared with user files    
+        /private
+            /keystore - keystore consists user's private key
+            /files/SIV - location of private files. SIV is 3 symbol path encryption algorithm identifier.
+```                   
+Example private profile:
+```
+    {
+        "keystore": {
+            "resource": "s3://adorsys-docusafe/users/username/private/keystore"
+        },
+        "privateStorage": [
+            [{"id": "DEFAULT"}, {"resource": "s3://adorsys-docusafe/users/username/private/files/"}]
+        ],
+        "inboxWithFullAccess": {
+            "resource": "s3://adorsys-docusafe/users/username/public/inbox/"
+        },
+        "publishPublicKeysTo": {
+            "resource": "s3://adorsys-docusafe/users/username/public/pubkeys"
+        },
+        "associatedResources": [
+            {"resource": "s3://adorsys-docusafe/users/username/"}
+        ],
+        "documentVersionStorage": {
+            "resource": "s3://adorsys-docusafe/users/username/versions/"
+        },
+        "appVersion": "BASELINE"
+    }
+```
+            
+Example public profile:
+```
+    {
+        "publicKeys": {
+            "resource": "s3://adorsys-docusafe/users/username/public/pubkeys"
+        },
+        "inbox": {
+            "resource": "s3://adorsys-docusafe/users/username/public/inbox/"
+        },
+        "appVersion": "BASELINE"
+    }   
+```
+    
+## Keystore encryption
```

Comment:
We use BCKFS to store keystore on disk and we use UBER to cache keys in memory (not yet but @electronicpeter is working on it)